File: Arrays/two_sum/TS/twoSum.ts
export {}

function twoSum(nums: number[], target: number): number[]{
    let differenceMap = new Map<number, number>();

    for(let i = 0; i <= nums.length; i++){
        const difference = target - nums[i];
        const validIdx = differenceMap.get(difference)
        if(validIdx !== undefined){
            return [validIdx, i]
        }
        differenceMap.set(nums[i], i)
    }

    return [-1]
}

const testCases: [number[], number][] = [
    [[2, 7, 11, 15], 9],
    [[3, 2, 4], 6],
    [[3, 3], 6],
    [[1, 2, 3, 4], 7],
    [[-1, -2, -3, -4], -6],
];

for (const [nums, target] of testCases) {
    console.log(`Input:       nums=${JSON.stringify(nums)}, target=${target}`);
    console.log(`Output:      ${JSON.stringify(twoSum(nums, target))}`);
    console.log("-".repeat(35));
}

File: Arrays/two_sum/python/two_sum.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        #difference(num) -> index
        difference_map = {}
        for idx, num in enumerate(nums):
            diff = target - num
            valid_idx = difference_map.get(diff, None)
            if valid_idx is not None:
                return [valid_idx, idx]
            difference_map[num] = idx
        return [-1]


if __name__ == "__main__":
    test_cases = [
        ([2, 7, 11, 15], 9),    # → [0, 1], classic example
        ([3, 2, 4], 6),         # → [1, 2], answer not at start
        ([3, 3], 6),            # → [0, 1], duplicate values
        ([1, 2, 3, 4], 7),      # → [2, 3], answer at end
        ([-1, -2, -3, -4], -6), # → [1, 3], negative numbers
    ]

    for nums, target in test_cases:
        print(f"Input:       nums={nums}, target={target}")
        print(f"Output:      {Solution().twoSum(nums, target)}")
        print("-" * 35)